test/wallet.test.ts

```diff
@@ -0,0 +1,339 @@
+      const { engine } = createTestSetup();
+      const getAccounts = async () => testAddresses.slice();
+      const witnessedMsgParams: MessageParams[] = [];
+      const processTypedMessage = async (msgParams: MessageParams) => {
+        witnessedMsgParams.push(msgParams);
+        return testTxHash;
+      };
+
+      engine.push(createWalletMiddleware({ getAccounts, processTypedMessage }));
+      const message = [
+        {
+          type: 'string',
+          name: 'message',
+          value: 'Hi, Alice!',
+        },
+      ];
+
+      const payload = {
+        method: 'eth_signTypedData',
+        params: [message, testUnkownAddress],
+      };
+      try {
+        await pify(engine.handle).call(engine, payload);
+      } catch (e: any) {
+        expect(e.message).toEqual(
+          'Invalid parameters: must provide an Ethereum address.',
+        );
+      }
+    });
+  });
+
+  describe('sign', () => {
+    it('valid address for sign', async () => {
+      const { engine } = createTestSetup();
+      const getAccounts = async () => testAddresses.slice();
+      const witnessedMsgParams: MessageParams[] = [];
+      const processPersonalMessage = async (msgParams: MessageParams) => {
+        witnessedMsgParams.push(msgParams);
+        return testMsgSig;
+      };
+
+      engine.push(
+        createWalletMiddleware({ getAccounts, processPersonalMessage }),
+      );
+
+      const message = 'haay wuurl';
+      const payload = {
+        method: 'personal_sign',
+        params: [message, testAddresses[0]],
+      };
+      const signMsgResponse = await pify(engine.handle).call(engine, payload);
+      const signMsgResult = signMsgResponse.result;
+
+      expect(signMsgResult).toBeDefined();
+      expect(signMsgResult).toEqual(testMsgSig);
+      expect(witnessedMsgParams.length).toEqual(1);
+      expect(witnessedMsgParams[0]).toStrictEqual({
+        data: message,
+        from: testAddresses[0],
+      });
+    });
+
+    it('invalid address for sign', async () => {
+      const { engine } = createTestSetup();
+      const getAccounts = async () => testAddresses.slice();
+      const witnessedMsgParams: MessageParams[] = [];
+      const processPersonalMessage = async (msgParams: MessageParams) => {
+        witnessedMsgParams.push(msgParams);
+        return testMsgSig;
+      };
+
+      engine.push(
+        createWalletMiddleware({ getAccounts, processPersonalMessage }),
+      );
+
+      const message = 'haay wuurl';
+      const payload = {
+        method: 'personal_sign',
+        params: [message, testUnkownAddress],
+      };
+
+      try {
+        await pify(engine.handle).call(engine, payload);
+      } catch (e: any) {
+        expect(e.message).toEqual(
+          'Invalid parameters: must provide an Ethereum address.',
+        );
+      }
+    });
+  });
+
+  describe('personalRecover', () => {
+    it('geth kumavis manual recover', async () => {
```

Comment:
Hmm well this one may be difficult to describe, what is this even doing exactly 😆